File: src/data/courses/dsa/trie/trie-design-variations/compressed-trie/code/python/compressed_trie.py
# ============================================================
# Solution 1: Brute Force — Hash Set / Direct comparison
# Time: O(N*L) or O(N²*L) | Space: O(N*L)
# ============================================================
class Solution1:
    def __init__(self): self.words = set()
    def insert(self, word): self.words.add(word)
    def hasPrefix(self, prefix):
        return any(w.startswith(prefix) for w in self.words)

# ============================================================
# Solution 1: Brute Force
# ============================================================

class Solution1:
    # Brute force: hash set / nested loops / direct comparison
    # See Solution 2 below for optimal Trie-based approach
    pass


# ============================================================
# Solution 2: Optimal (Trie-based)
# ============================================================
class RadixNode:
    def __init__(self):
        self.children = {}
        self.is_end = False
        self.edge_label = ""

class CompressedTrie:
    def __init__(self):
        self.root = RadixNode()
    
    def insert(self, word):
        node = self.root
        i = 0
        while i < len(word):
            char = word[i]
            if char not in node.children:
                new_node = RadixNode()
                new_node.edge_label = word[i:]
                new_node.is_end = True
                node.children[char] = new_node
                return
            
            child = node.children[char]
            label = child.edge_label
            j = 0
            while j < len(label) and i < len(word) and label[j] == word[i]:
                i += 1
                j += 1
            
            if j < len(label):
                # Split edge
                split_node = RadixNode()
                split_node.edge_label = label[:j]
                split_node.children[label[j]] = child
                child.edge_label = label[j:]
                node.children[char] = split_node
                
                if i < len(word):
                    new_node = RadixNode()
                    new_node.edge_label = word[i:]
                    new_node.is_end = True
                    split_node.children[word[i]] = new_node
                else:
                    split_node.is_end = True
                return
            node = child
        node.is_end = True
    
    def search(self, word):
        node = self.root
        i = 0
        while i < len(word):
            char = word[i]
            if char not in node.children:
                return False
            child = node.children[char]
            label = child.edge_label
            for c in label:
                if i >= len(word) or word[i] != c:
                    return False
                i += 1
            node = child
        return node.is_end

File: src/data/courses/dsa/trie/trie-design-variations/compressed-trie/code/java/CompressedTrie.java
// ============================================================
// Solution 1: Brute Force
// ============================================================
import java.util.*;

class Solution1 {
    // Brute force: hash set / nested loops / direct comparison
    // See Solution 2 below for optimal Trie-based approach
}

// ============================================================
// Solution 2: Optimal (Trie-based)
// ============================================================
import java.util.*;

// Time: O(L) all operations | Space: O(N) nodes

class RadixNode {
    Map<Character, RadixNode> children = new HashMap<>();
    boolean isEnd = false;
    String edgeLabel = "";
}

class CompressedTrie {
    private RadixNode root;
    
    public CompressedTrie() {
        root = new RadixNode();
    }
    
    public void insert(String word) {
        RadixNode node = root;
        int i = 0;
        
        while (i < word.length()) {
            char ch = word.charAt(i);
            if (!node.children.containsKey(ch)) {
                RadixNode newNode = new RadixNode();
                newNode.edgeLabel = word.substring(i);
                newNode.isEnd = true;
                node.children.put(ch, newNode);
                return;
            }
            
            RadixNode child = node.children.get(ch);
            String label = child.edgeLabel;
            int j = 0;
            
            while (j < label.length() && i < word.length() && label.charAt(j) == word.charAt(i)) {
                i++; j++;
            }
            
            if (j < label.length()) {
                // Split edge
                RadixNode splitNode = new RadixNode();
                splitNode.edgeLabel = label.substring(0, j);
                splitNode.children.put(label.charAt(j), child);
                child.edgeLabel = label.substring(j);
                node.children.put(ch, splitNode);
                
                if (i < word.length()) {
                    RadixNode newNode = new RadixNode();
                    newNode.edgeLabel = word.substring(i);
                    newNode.isEnd = true;
                    splitNode.children.put(word.charAt(i), newNode);
                } else {
                    splitNode.isEnd = true;
                }
                return;
            }
            node = child;
        }
        node.isEnd = true;
    }
    
    public boolean search(String word) {
        RadixNode node = root;
        int i = 0;
        
        while (i < word.length()) {
            char ch = word.charAt(i);
            if (!node.children.containsKey(ch)) return false;
            
            RadixNode child = node.children.get(ch);
            String label = child.edgeLabel;
            
            for (char c : label.toCharArray()) {
                if (i >= word.length() || word.charAt(i) != c) return false;
                i++;
            }
            node = child;
        }
        return node.isEnd;
    }
}
